packages/mcfunction/src/parser/macro.ts

```diff
@@ -0,0 +1,122 @@
+import * as core from '@spyglassmc/core'
+import { localeQuote, localize } from '@spyglassmc/locales'
+import type {
+	MacroGapNode,
+	MacroKeyNode,
+	MacroNode,
+	MacroSignNode,
+	MacroVariableNode,
+} from '../node/index.js'
+
+/**
+ * Parse a macro line.
+ */
+export function macro(): core.Parser<MacroNode> {
+	return (
+		src: core.Source,
+		ctx: core.ParserContext,
+	): core.Result<MacroNode> => {
+		const children:
+			(MacroGapNode | MacroVariableNode | MacroKeyNode | MacroSignNode)[] =
+				[]
+
+		// Skip the starting '$'
+		children.push({
+			type: 'mcfunction:macro/sign',
+			range: core.Range.create(src.cursor, src.cursor + 1),
+			value: '$',
+			path: ['$'],
+		})
+		src.skip()
+
+		let start = src.cursor
+		let hasMacroKeys = false
+
+		while (src.canReadInLine()) {
+			src.skipUntilOrEnd(core.LF, core.CR, '$')
+			if (src.peek(2) === '$(') {
+				// Add the gap before this macro key
+				const gap = src.sliceToCursor(start)
+				if (gap.length > 0) {
+					children.push({
+						type: 'mcfunction:macro/gap',
+						range: core.Range.create(start, src.cursor),
+						value: gap,
+						path: [gap],
+					})
+					start = src.cursor
+				}
+				// Parse the macro key
+				src.skip(2)
+				const keyStart = src.cursor
+				src.skipUntilOrEnd(core.LF, core.CR, ')')
+				if (src.peek() != ')') {
+					// Macro key was not closed
+					ctx.err.report(
+						localize('expected', localeQuote(')')),
+						core.Range.create(keyStart, src.cursor),
+					)
+				} else if (src.cursor <= keyStart) {
+					// Encountered $()
+					ctx.err.report(
+						localize('expected', localize('macro-key')),
```

Comment:
No change to en.json, so presumably missing translation for `macro-key`

Reply:
There's other missing English locale strings too, should probably do them all in-batch. We also need an issue for it

Reply:
> There's other missing English locale strings too, should probably do them all in-batch.

i kinda strongly disagree with this, if we're actively referencing a new locale string in a PR we should be diligent about adding it

(easier to do this with people being active with reviews)

Reply:
alright, well looks like Calverin added them anyway